rich/progress.py

```diff
@@ -149,6 +168,321 @@ def track(
         )
 
 
+class _Reader(RawIOBase, BinaryIO):
+    """A reader that tracks progress while it's being read from."""
+
+    def __init__(
+        self,
+        handle: BinaryIO,
+        progress: "Progress",
+        task: TaskID,
+        close_handle: bool = True,
+    ) -> None:
+        self.handle = handle
+        self.progress = progress
+        self.task = task
+        self.close_handle = close_handle
+        self._closed = False
+
+    def __enter__(self) -> "_Reader":
+        self.handle.__enter__()
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_val: Optional[BaseException],
+        exc_tb: Optional[TracebackType],
+    ) -> None:
+        self.close()
+
+    def __iter__(self) -> BinaryIO:
+        return self
+
+    def __next__(self) -> bytes:
+        line = next(self.handle)
+        self.progress.advance(self.task, advance=len(line))
+        return line
+
+    @property
+    def closed(self) -> bool:
+        return self._closed
+
+    def fileno(self) -> int:
+        return self.handle.fileno()
+
+    def isatty(self) -> bool:
+        return self.handle.isatty()
+
+    def readable(self) -> bool:
+        return self.handle.readable()
+
+    def seekable(self) -> bool:
+        return self.handle.seekable()
+
+    def writable(self) -> bool:
+        return False
+
+    def read(self, size: int = -1) -> bytes:
+        block = self.handle.read(size)
+        self.progress.advance(self.task, advance=len(block))
+        return block
+
+    def readall(self) -> bytes:
+        block = self.handle.readall()  # type: ignore
+        self.progress.advance(self.task, advance=len(block))
+        return block  # type: ignore
+
+    def readinto(self, b: Union[bytearray, memoryview, mmap]):  # type: ignore
+        n = self.handle.readinto(b)  # type: ignore
+        self.progress.advance(self.task, advance=n)
+        return n
+
+    def readline(self, size: int = -1) -> bytes:  # type: ignore
```

Comment:
What's the thinking behind these `type: ignore` comments?

As far as I can see,`self.handle` which is typed as a `BinaryIO` doesn't contain `readinto` or `readall` (those come from `RawIOBase`). `mypy`s warnings seem valid to me here.

Indeed, calling `readall` on the `_Reader` instance will result in `AttributeError: '_io.BufferedReader' object has no attribute 'readall'. Did you mean: 'readable'?`

I'm wondering if these methods are required?

Reply:
So as far as I've seen while doing I/O related typing in Pyfilesystem2, the I/O typing is really inconsistent between the `typing` and the `io` modules.

Basically, when you open a file in binary mode with `open`, you get either an `io.BufferedReader` or an `io.FileIO` (if buffering is disabled), both of which have a `readinto` method, . However when `mypy` cannot determine the buffering statically, it will have `open` default to returning a `BinaryIO`, which doesn't have these methods, although the actual objects being returned will always have `readinto`:

https://github.com/python/typeshed/blob/9687d53b65d055a990f0187b7baa1137cb7c717b/stdlib/builtins.pyi#L1331-L1341

That's why I decided to silence `mypy` here. I can definitely remove the `readall` implementation, but I'd leave the `readinto`, in particular because it can really improve performance in some cases by avoiding useless copies of byte data.

Reply:
Will defer to your experience here, Martin! Regarding the `readall`, it is probably better to be defensive here even if it defies the typing.

Reply:
I think it's actually safer here to remove `readall` because I'll expect the majority of calls to `progress.open` to be buffered, so the returned `io.BufferedReader` won't have a `readall` method. It's also much more marginal than `readinto` in my experience, so I don't think you'll get a lot complaints that `_Reader` misses a `readall` method because there's very limited reason to use it instead of `read()` without arguments.